### PR TITLE
6. As a user, I want to be alerted when I’m entering an item that is the same as something already on my list so I can avoid duplicate items. 

### LINKED ISSUE #6: 6. As a user, I want to be alerted when I’m entering an item that is the same as something already on my list so I can avoid duplicate items.
AC:

* Show an error message if the user tries to submit a new item that has the exact same name as an existing item
* Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
* The user’s original input should be what gets saved in the database

### PR DESCRIPTION
**Pull request by jessicasalbert**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

Previous behavior allowed a user to enter anything in the form, including a blank value. This PR implements basic form validation to require a user to input a value for all form elements using the 'required' attribute. 

The app also allowed for duplicate entries. This PR implements a check upon user submission of a new shopping list item to see if that item is already present in the database. The database stores the user's initial entry unchanged, then any subsequent item is compared with capitalization normalized and punctuation removed to avoid duplicates. 

One issue we faced was in dealing with asynchronicity. Our function to get items from the database and compare for duplicates was being performed asynchronously, which was throwing off the order of execution and our duplicate check wasn't working correctly. We used async/await in two functions to ensure that they were executed in order. 

## Related Issue

Closes #6 

## Acceptance Criteria

AC:

* Show an error message if the user tries to submit a new item that has the exact same name as an existing item
* Show an error message if the user tries to submit a new item that has the same name as an existing item, where  capitalization has been normalized and punctuation has been removed
* The user’s original input should be what gets saved in the database

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### Before

image

### After

User is alerted when item entered is a duplicate

[Image from Gyazo](https://gyazo.com/894118805e6626fa1aea09843d052921)

## Testing Steps / QA Criteria

* Clone, run npm install to update dependencies, then run npm start to run app
* Submit "add item" form with no input - should be alerted that name is required
* Submit item with no frequency selected - should be alerted that frequency is required
* Enter an item that is already on the list. Use different spelling & punctuation - should still be alerted that item is a duplicate
* Duplicate items should not appear in list view

### CODE REVIEW

File: src/components/AddItem.js [hunk 1]
```diff
@@ -12,8 +12,13 @@ const AddItem = () => {
     setGroceryItem(e.target.value);
   };
 
-  const onSubmitHandler = (e) => {
+  async function onSubmitHandler(e) {
```
Comment (Maeesha-Rahman) [inline src/components/AddItem.js]: excellent use of async and await! :fire:

File: src/components/AddItem.js [hunk 2]
```diff
@@ -24,6 +29,30 @@ const AddItem = () => {
...
+          ) {
+            alreadyExists = true;
+          }
+        });
+      })
+      .catch((e) => {
+        alert(e.message);
+      });
+    return alreadyExists;
+  };
```
Comment (Maeesha-Rahman) [inline src/components/AddItem.js]: love the error handling! :clap:

**Review by Maeesha-Rahman — APPROVED** [review]
Amazing job you two! Looks good on my end! Nice work using regex!! :sparkles:

File: src/components/AddItem.js [hunk 3]
```diff
@@ -40,6 +69,7 @@ const AddItem = () => {
           type="text"
           value={groceryItem}
           onChange={onGroceryItemInputChange}
+          required
```
Comment (wlcreate) [inline src/components/AddItem.js:72]: I love that you took into consideration the edge case of a user not inputting anything and submitting the form by adding "required" to all of the inputs.

**Review by wlcreate — APPROVED** [review]
Awesome job Kevin and Jessica! I love your use of async/await to deal with the order of execution and safety check of a user adding an item without any input. I was a bit confused with what ```sanitizeItemInput``` was at first; once I realized what it was doing I understood why you chose to name it that. 

My only suggestion is to move ```onRadioInputChange``` and ```sanitizeItemInput``` to be between ```onGroceryItemInputChange``` and ```onSubmitHandler``` since they are smaller functions.

**Review by MelinaMejia95 — CHANGES_REQUESTED** [review]
This looks amazing!!! Super great work! 🎊
I just have general feedback:

- In some parts of the code, you're using "" and '' in some other places. To keep the code cleaner, you should talk as a team and discuss one of those you're going to use 😺
- Also, to keep the code cleaner, as you're using const someCoolName = () => {} structure in most of the methods you're creating, you should keep that consistency, I see a lot of function () {}, it would be nice to have all methods with the same structure ❤️

**Review by MelinaMejia95 — APPROVED** [review]

**Comment by thekidnamedkd** [comment]
Yahooooo

**Review by skylerwshaw — APPROVED** [review]
Good use of regex! Overall 💯 !

**Pull request merged**